src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java

```diff
@@ -186,41 +189,64 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
             firestore, collectionId, DELETE_BATCH_SIZE, document -> document.getReference().delete());
     }
 
+
+    interface LookupFunction {
+        DocumentSnapshot apply(Transaction xn) throws InterruptedException;
+    }
+
+    // Returns null if not found - upper layers do any throwing
+    private FireStoreDirectoryEntry runTransactionWithRetry(Firestore firestore,
+                                                            LookupFunction lookupFunction,
+                                                            String transactionOp,
+                                                            String warnMessage) throws InterruptedException {
+        for (int i = 0; i < RETRIES; i++) {
+            try {
+                ApiFuture<FireStoreDirectoryEntry> transaction =
+                    firestore.runTransaction(
+                        xn -> {
+                            DocumentSnapshot docSnap = lookupFunction.apply(xn);
+                            if (docSnap == null) {
+                                return null;
+                            }
+                            return docSnap.toObject(FireStoreDirectoryEntry.class);
+                        });
+
+                return fireStoreUtils.transactionGet(transactionOp, transaction);
+            } catch (Exception ex) {
+                if (FireStoreUtils.shouldRetry(ex)) {
+                    logger.warn("Retry-able error in firestore future get - " +
+                        warnMessage + " message: " + ex.getMessage());
+                } else {
+                    throw new FileSystemExecutionException(transactionOp + " execution exception", ex);
+                }
+            }
+
+            TimeUnit.MILLISECONDS.sleep(SLEEP_MILLISECONDS);
```

Comment:
It's a totally minor thing but if you're using `TimeUnit` anyway it would be more natural IMO to write
```suggestion
            TimeUnit.SECONDS.sleep(1);
```

Reply:
I started using MILLISECONDS for the FireStore waits, because it is likely that we will want to turn these down to sub-second values. 1000ms works, but I bet we could get by with 250. It is a case of, get it working first and tune it up later.